src/arangodb.c

```diff
@@ -0,0 +1,1201 @@
+
+
+//
+// service is one ArangoDB component:  agent, coordinator, or db server.
+//   service_t objects are created dynamically after reading cluster configuration
+//   via cluster_t's endpoint
+//
+struct service_s {
+  endpoint_t endpoint;
+  char * service_name;
+  char * registered_name;
+  char * role;
+  char * engine;
+  bool engine_init_done;
+  curlinfo_t curlinfo;
+};
+
+typedef struct service_s service_t;
+
+//
+// callback declares
+//
+static int arangodb_complex_config(oconfig_item_t * config_item);
+static int arangodb_init(void);
+
+static int cluster_read(user_data_t *ud);
+static size_t cluster_write_callback(void *buf,
+                                     size_t size, size_t nmemb,
+                                     void *user_data);
+
+
+// support functions
+static cluster_t * new_cluster(void);
+static void free_cluster(void * arg);
+static void free_endpoint(endpoint_t * endpoint);
+static void free_curlinfo(curlinfo_t * curlinfo);
+static int arangodb_add_cluster(oconfig_item_t * config_item);
+static int cluster_set_options(oconfig_item_t * config_item, cluster_t * cluster);
+static int cluster_verify_options(cluster_t * cluster);
+static void cluster_decode_health(yajl_val health_node);
+
+static void start_curl_session(curlinfo_t * curlinfo);
+static int curl_perform(curlinfo_t * curlinfo, endpoint_t * endpoint, char * page, yajl_val * response);
+
+static service_t * new_service_single(cluster_t * cluster);
+static service_t * new_service_cluster(yajl_val health_node, const char * service_name);
+static void free_service(void * arg);
+static int service_read(user_data_t *ud);
+static void service_decode_stats(service_t * service, yajl_val srv_stats, yajl_val stats_desc);
+static void service_data_set_init(yajl_val figures_array);
+static void service_get_engine(service_t * service);
+static int service_get_rocksdb(service_t * service);
+
+static char * arangodb_get_role(curlinfo_t * curlinfo, endpoint_t * endpoint);
+static bool rocksdb_is_gauge(const char * key);
+static bool endpoint_parse(endpoint_t * endpoint);
+
+// management for first time init of server data_set
+static pthread_mutex_t service_data_set_mutex = PTHREAD_MUTEX_INITIALIZER;
+static bool service_data_set_done = false;
+
+static pthread_mutex_t rocks_data_set_mutex = PTHREAD_MUTEX_INITIALIZER;
+static bool rocks_data_set_done = false;
+
+
+//
+// entry point from collectd daemon
+//
+void module_register(void) {
+  plugin_register_complex_config("arangodb", arangodb_complex_config);
+  plugin_register_init("arangodb", arangodb_init);
+} /* void module_register */
+
+
+static int arangodb_init(void) {
+
+  // taken from curl_json.c
+  /* Call this while collectd is still single-threaded to avoid
+   * initialization issues in libgcrypt. */
+  curl_global_init(CURL_GLOBAL_ALL);
+
+  return 0;
+} // arangodb_init
+
+
+//
+// Config level one:  cluster / singleserver key word
+//
+static int arangodb_complex_config(oconfig_item_t * config_item) {
```

Comment:
Typically, order of functions in code is upwards by call order.
Then you do not need to do forward  declarations.

So, preferred order looks like

... and so on.
arangodb_add_cluster
arangodb_complex_config
arangodb_init
module_register